File: _inc/_DEPRECATED_django/erp_prestech_backend/app/Http/Controllers/charts/chart_of_account_controller.py
import calendar  
import datetime  
from django.contrib import messages  
from django.http import HttpRequest, HttpResponse, JsonResponse  
from django.shortcuts import get_object_or_404, redirect, render  
from ....Models.charts.chart_of_account import ChartOfAccount
from ....Models.charts.chart_of_account_sub_type import ChartOfAccountSubType
from ....Models.charts.chart_of_account_type import ChartOfAccountType
from .._traits.controller import Controller
from .._helpers.http import get_redirect_url
from django.core.exceptions import PermissionDenied
from ....configs.messages_templates import get_exception_class_message
import logging  
logger = logging.getLogger(__name__)  

class ChartOfAccountController(Controller):  
  def index(self, request: HttpRequest) -> HttpResponse:  
    try:  
      if request.user.has_perm('manage chart of account'):  
        start = request.GET.get('start_date') if request.GET.get('start_date') and request.GET.get('end_date') else datetime.date.today().replace(month=1, day=1).strftime('%Y-%m-%d')  
        end = request.GET.get('end_date') if request.GET.get('start_date') and request.GET.get('end_date') else (datetime.date.today() + datetime.timedelta(days=1)).strftime('%Y-%m-%d')  
        filter = {'startDateRange': start, 'endDateRange': end}  
        types = ChartOfAccountType.objects.filter(created_by=request.user.creator_id())  
        chartAccounts = {}  
        for t in types:  
          accounts = ChartOfAccount.objects.filter(type=t.id, created_by=request.user.creator_id()).select_related('subType')  
          chartAccounts[t.name] = accounts  
        logger.info("ChartOfAccountController.index: Loaded successfully")  
        return render(request, 'chartOfAccount/index.html', {'chartAccounts': chartAccounts, 'types': types, 'filter': filter})  
      else:  
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))  
        return redirect(get_redirect_url(request))  
    except Exception as e:  
      logger.exception(f"ChartOfAccountController.index failed: {e}")  
      messages.error(request, "An error occurred.")  
      return redirect(get_redirect_url(request))  
  
  def create(self, request: HttpRequest) -> HttpResponse:  
    try:  
      types_qs = ChartOfAccountType.objects.filter(created_by=request.user.creator_id()).values('id', 'name')  
      types_dict = {t['id']: t['name'] for t in types_qs}  
      types_dict = {0: 'Select Account Type', **types_dict}  
      logger.info("ChartOfAccountController.create: Rendered create view")  
      return render(request, 'chartOfAccount/create.html', {'types': types_dict})  
    except Exception as e:  
      logger.exception(f"ChartOfAccountController.create failed: {e}")  
      messages.error(request, "An error occurred.")  
      return redirect(get_redirect_url(request))  
  
  def store(self, request: HttpRequest) -> HttpResponse:  
    try:  
      if request.user.has_perm('create chart of account'):  
        name = request.POST.get('name')  
        type_val = request.POST.get('type')  
        if not name or not type_val:  
          messages.error(request, "Name and Account Type are required.")  
          return redirect(get_redirect_url(request))  
        account = ChartOfAccount()  
        account.name = name  
        account.code = request.POST.get('code')  
        account.type = type_val  
        account.sub_type = request.POST.get('sub_type')  
        account.description = request.POST.get('description')  
        account.is_enabled = 1 if request.POST.get('is_enabled') else 0  
        account.created_by = request.user.creator_id()  
        account.save()  
        messages.success(request, "Account successfully created.")  
        logger.info("ChartOfAccountController.store: Account created")  
        return redirect('chart-of-account_index')  
      else:  
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))  
        return redirect(get_redirect_url(request))  
    except Exception as e:  
      logger.exception(f"ChartOfAccountController.store failed: {e}")  
      messages.error(request, f"ChartOfAccountController.store: {e}")  
      return redirect(get_redirect_url(request))  
  
  def show(self, request: HttpRequest, id: int) -> HttpResponse:  
    try:  
      if request.user.has_perm('ledger report'):  
        start = request.GET.get('start_date') if request.GET.get('start_date') and request.GET.get('end_date') else datetime.date.today().replace(day=1).strftime('%Y-%m-%d')  
        last_day = calendar.monthrange(datetime.date.today().year, datetime.date.today().month)[1]  
        end = request.GET.get('end_date') if request.GET.get('start_date') and request.GET.get('end_date') else datetime.date.today().replace(day=last_day).strftime('%Y-%m-%d')  
        accounts_qs = ChartOfAccount.objects.filter(created_by=request.user.creator_id())  
        accounts_qs = accounts_qs.filter(created_at__gte=start, created_at__lte=end) if request.GET.get('start_date') and request.GET.get('end_date') else accounts_qs  
        accounts = {}  
        for acc in accounts_qs:  
          accounts[acc.id] = f"{acc.code} - {acc.name}"  
        accounts = {'': 'Select Account', **accounts}  
        account_id = request.GET.get('account')  
        account = ChartOfAccount.objects.filter(pk=account_id).first() if account_id else ChartOfAccount.objects.filter(pk=id).first()  
        filter = {'startDateRange': start, 'endDateRange': end}  
        logger.info("ChartOfAccountController.show: Rendered show view")  
        return render(request, 'chartOfAccount/show.html', {'filter': filter, 'account': account, 'accounts': accounts})  
      else:  
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))  
        return redirect(get_redirect_url(request))  
    except Exception as e:  
      logger.exception(f"ChartOfAccountController.show failed: {e}")  
      messages.error(request, "An error occurred.")  
      return redirect(get_redirect_url(request))  
  
  def edit(self, request: HttpRequest, id: int) -> HttpResponse:  
    try:  
      chartOfAccount = get_object_or_404(ChartOfAccount, pk=id)  
      types_qs = ChartOfAccountType.objects.all().values('id', 'name')  
      types_dict = {t['id']: t['name'] for t in types_qs}  
      types_dict = {0: 'Select Account Type', **types_dict}  
      logger.info("ChartOfAccountController.edit: Rendered edit view for ID %s", id)  
      return render(request, 'chartOfAccount/edit.html', {'chartOfAccount': chartOfAccount, 'types': types_dict})  
    except Exception as e:  
      logger.exception(f"ChartOfAccountController.edit failed: {e}")  
      messages.error(request, "An error occurred.")  
      return redirect(get_redirect_url(request))  
  
  def update(self, request: HttpRequest, id: int) -> HttpResponse:  
    try:  
      if request.user.has_perm('edit chart of account'):  
        chartOfAccount = get_object_or_404(ChartOfAccount, pk=id)  
        name = request.POST.get('name')  
        if not name:  
          messages.error(request, "Name is required.")  
          return redirect(get_redirect_url(request))  
        chartOfAccount.name = name  
        chartOfAccount.code = request.POST.get('code')  
        chartOfAccount.description = request.POST.get('description')  
        chartOfAccount.is_enabled = 1 if request.POST.get('is_enabled') else 0  
        chartOfAccount.save()  
        messages.success(request, "Account successfully updated.")  
        logger.info("ChartOfAccountController.update: Account ID %s updated", id)  
        return redirect('chart-of-account_index')  
      else:  
        return JsonResponse({'error': "Permission denied."}, status=401)  
    except Exception as e:  
      logger.exception(f"ChartOfAccountController.update failed: {e}")  
      messages.error(request, f"ChartOfAccountController.update: {e}")  
      return redirect(get_redirect_url(request))  
  
  def destroy(self, request: HttpRequest, id: int) -> HttpResponse:  
    try:  
      chartOfAccount = get_object_or_404(ChartOfAccount, pk=id)  
      if request.user.has_perm('delete chart of account'):  
        chartOfAccount.delete()  
        messages.success(request, "Account successfully deleted.")  
        logger.info("ChartOfAccountController.destroy: Account ID %s deleted", id)  
        return redirect('chart-of-account_index')  
      else:  
        messages.error(request,get_exception_class_message(PermissionDenied, __class__.__name__))  
        return redirect(get_redirect_url(request))  
    except Exception as e:  
      logger.exception(f"ChartOfAccountController.destroy failed: {e}")  
      messages.error(request, "An error occurred.")  
      return redirect(get_redirect_url(request))  
  
  def get_subtype(self, request: HttpRequest) -> HttpResponse:  
    try:  
      type_val = request.POST.get('type')  
      types_qs = ChartOfAccountSubType.objects.filter(type=type_val).values('id', 'name')  
      types_dict = {t['id']: t['name'] for t in types_qs}  
      logger.info("ChartOfAccountController.get_subtype: Returned subtypes for type %s", type_val)  
      return JsonResponse(types_dict)  
    except Exception as e:  
      logger.exception(f"ChartOfAccountController.get_subtype failed: {e}")  
      return JsonResponse({'error': "An error occurred."}, status=500)  

File: _inc/laravel/app/Http/Controllers/Charts/ChartOfAccountController.php
<?php

namespace App\Http\Controllers;

use App\Config\Constants\{
    ChartsConstants,
    DatabaseConstants,
    PermissionsConstants,
    UsersConstants,
};
use App\Models\{
    ChartOfAccount,
    ChartOfAccountSubType,
    ChartOfAccountType,
    JournalItem
};
use App\Traits\{ChecksLogin, ChecksPermissions};
use Illuminate\Http\{JsonResponse, RedirectResponse, Request};
use Illuminate\Support\Facades\{Validator, View as ViewFacade};
use Illuminate\View\View;
use Symfony\Component\HttpFoundation\Response;
use Throwable;

final class ChartOfAccountController extends Controller
{
    use ChecksLogin, ChecksPermissions;
    private const ROUTE_SINGULAR = 'chart-of-account'; // ! ALERT
    private const REDIRECT_INDEX = self::ROUTE_SINGULAR . '.index'; // ! ALERT

    public function index(Request $req): View|RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($req, $func, $action) {
            if (($uor = self::_checkLogin()) instanceof RedirectResponse) return $uor;
            $user = $uor;
            if (($c = self::guard($req, PermissionsConstants::MNG_COA, self::REDIRECT_INDEX)) !== true) return $c;

            try {
                [$start, $end] = $req->filled(['start_date', 'end_date'])
                    ? [$req->start_date, $req->end_date]
                    : [date('Y-01-01'), date('Y-m-d', strtotime('+1 day'))];

                $t = microtime(true);
                $types = ChartOfAccountType::whereCreatedBy($user?->creatorId())->get();
                $this->logExecutionTime($t, $action, 'loadTypes');

                $t2 = microtime(true);
                $chartAccounts = $types->mapWithKeys(fn($t) => [
                    $t->name => ChartOfAccount::where([
                        [ChartsConstants::COL_TP, $t->id],
                        [DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId()]
                    ])->with(ChartsConstants::COL_SUBTP)->get(),
                ]);
                $this->logExecutionTime($t2, $action, 'loadAccountsByType');

                $view = self::ROUTE_SINGULAR . '.' . $func;
                if (!ViewFacade::exists($view)) return defaultUndefinedException($req, new \Exception('view'), $action, route(self::REDIRECT_INDEX));

                return ViewFacade::make($view, [
                    'chartAccounts' => $chartAccounts,
                    'types'         => $types,
                    'filter'        => [
                        'startDateRange' => $start,
                        'endDateRange'   => $end,
                    ],
                ]);
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action, route(self::REDIRECT_INDEX));
            }
        }, [
            UsersConstants::COL_USER_ID => $req->user()?->id ?? null,
            'start_date' => $req->input('start_date'),
            'end_date' => $req->input('end_date'),
        ]);
    }

    public function create(Request $req): View|RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($req, $func, $action) {
            if (($uor = self::_checkLogin()) instanceof RedirectResponse) return $uor;
            $user = $uor;
            if (($c = self::guard($req, PermissionsConstants::CR_COA, self::REDIRECT_INDEX)) !== true) return $c;

            try {
                $t = microtime(true);
                $types = ChartOfAccountType::whereCreatedBy($user?->creatorId())
                    ->pluck(ChartsConstants::COL_NM, 'id')
                    ->prepend('Select Account Type', '');
                $this->logExecutionTime($t, $action, 'loadTypesForCreate');

                $view = self::ROUTE_SINGULAR . '.' . $func;
                if (!ViewFacade::exists($view)) return defaultUndefinedException($req, new \Exception('view'), $action, route(self::REDIRECT_INDEX));

                return ViewFacade::make($view, compact('types'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action, route(self::REDIRECT_INDEX));
            }
        }, [UsersConstants::COL_USER_ID => $req->user()?->id ?? null]);
    }

    public function store(Request $req): RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($req, $action) {
            if (($uor = self::_checkLogin()) instanceof RedirectResponse) return $uor;
            $user = $uor;
            if (($c = self::guard($req, PermissionsConstants::CR_COA, self::REDIRECT_INDEX)) !== true) return $c;
            if ($c = self::v($req, [
                ChartsConstants::COL_NM    => 'required|string|max:255',
                ChartsConstants::COL_CD    => 'required|integer',
                ChartsConstants::COL_TP    => 'required|uuid|exists:chart_of_account_types,id',
                ChartsConstants::COL_SUBTP => 'required|uuid|exists:chart_of_account_sub_types,id',
            ])) return $c;

            try {
                $t = microtime(true);
                ChartOfAccount::create([
                    ChartsConstants::COL_NM    => $req->name,
                    ChartsConstants::COL_CD    => $req->code,
                    ChartsConstants::COL_TP    => $req->type,
                    ChartsConstants::COL_SUBTP => $req->sub_type,
                    ChartsConstants::COL_DESC  => $req->description,
                    ChartsConstants::COL_ENB   => $req->has(ChartsConstants::COL_ENB) ? 1 : 0,
                    DatabaseConstants::COL_TABLE_CREATOR => $user?->creatorId(),
                ]);
                $this->logExecutionTime($t, $action, 'createAccount');

                return redirect()->route(self::REDIRECT_INDEX)->with('success', __('Account successfully created.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action, route(self::REDIRECT_INDEX));
            }
        }, [
            UsersConstants::COL_USER_ID => $req->user()?->id ?? null,
            'input' => $req->only('name', 'code', 'type', 'sub_type'),
        ]);
    }

    public function show(ChartOfAccount $chartOfAccount, Request $req): View|RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($req, $chartOfAccount, $func, $action) {
            if (($uor = self::_checkLogin()) instanceof RedirectResponse) return $uor;
            $user = $uor;
            if (($c = self::guard($req, PermissionsConstants::LDG_RPT, self::REDIRECT_INDEX)) !== true) return $c;

            try {
                [$start, $end] = $req->filled(['start_date', 'end_date'])
                    ? [$req->start_date, $req->end_date]
                    : [date('Y-m-01'), date('Y-m-t')];

                $colCd = ChartsConstants::COL_CD;
                $colNm = ChartsConstants::COL_NM;

                $t = microtime(true);
                $accounts = ChartOfAccount::selectRaw("CONCAT({$colCd}, ' - ', {$colNm}) AS code_name, id")
                    ->whereCreatedBy($user?->creatorId())
                    ->when($req->filled(['start_date', 'end_date']), fn($q) => $q->whereBetween('created_at', [$start, $end]))
                    ->pluck('code_name', 'id')
                    ->prepend('Select Account', '');
                $this->logExecutionTime($t, $action, 'loadAccountList');

                $account = $req->account ? ChartOfAccount::find($req->account) : $chartOfAccount;

                $t2 = microtime(true);
                $journalItems = JournalItem::select(
                    DatabaseConstants::TABLE_JOURNAL_ENTRIES . '.journal_id',
                    DatabaseConstants::TABLE_JOURNAL_ENTRIES . '.date AS transaction_date',
                    'journal_items.debit',
                    'journal_items.credit'
                )
                    ->join('journal_entries', DatabaseConstants::TABLE_JOURNAL_ENTRIES . '.id', '=', 'journal_items.journal')
                    ->where(DatabaseConstants::TABLE_JOURNAL_ENTRIES . '.' . DatabaseConstants::COL_TABLE_CREATOR, $user?->creatorId())
                    ->where('journal_items.account', $account->id)
                    ->whereBetween(DatabaseConstants::TABLE_JOURNAL_ENTRIES . '.date', [$start, $end])
                    ->get();
                $this->logExecutionTime($t2, $action, 'loadJournalItems');

                $debit  = $journalItems->sum('debit');
                $credit = $journalItems->sum('credit');
                $balance = $credit - $debit;

                $view = self::ROUTE_SINGULAR . '.' . $func;
                if (!ViewFacade::exists($view)) return defaultUndefinedException($req, new \Exception('view'), $action, route(self::REDIRECT_INDEX));

                return ViewFacade::make($view, [
                    'filter'       => [
                        'startDateRange' => $start,
                        'endDateRange'   => $end,
                    ],
                    'account'      => $account,
                    'accounts'     => $accounts,
                    'journalItems' => $journalItems,
                    'debit'        => $debit,
                    'credit'       => $credit,
                    'balance'      => $balance,
                ]);
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action, route(self::REDIRECT_INDEX));
            }
        }, [
            UsersConstants::COL_USER_ID => $req->user()?->id ?? null,
            'account_id' => $chartOfAccount->id ?? null,
        ]);
    }

    public function edit(Request $req, ChartOfAccount $chartOfAccount): View|RedirectResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($req, $chartOfAccount, $func, $action) {
            if (($uor = self::_checkLogin()) instanceof RedirectResponse) return $uor;
            if (($c = self::guard($req, 'edit chart of account', self::REDIRECT_INDEX)) !== true) return $c;

            try {
                $t = microtime(true);
                $types = ChartOfAccountType::pluck(ChartsConstants::COL_NM, 'id')->prepend('Select Account Type', '');
                $this->logExecutionTime($t, $action, 'loadTypesForEdit');

                $view = self::ROUTE_SINGULAR . '.' . $func;
                if (!ViewFacade::exists($view)) return defaultUndefinedException($req, new \Exception('view'), $action, route(self::REDIRECT_INDEX));

                return ViewFacade::make($view, compact('chartOfAccount', 'types'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action, route(self::REDIRECT_INDEX));
            }
        }, ['account_id' => $chartOfAccount->id ?? null]);
    }

    public function update(Request $req, ChartOfAccount $chartOfAccount): RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($req, $chartOfAccount, $action) {
            if (($uor = self::_checkLogin()) instanceof RedirectResponse) return $uor;
            if (($c = self::guard($req, 'edit chart of account', self::REDIRECT_INDEX)) !== true) return $c;
            if ($c = self::v($req, [ChartsConstants::COL_NM => 'required'])) return $c;

            try {
                $t = microtime(true);
                $chartOfAccount->update([
                    ChartsConstants::COL_NM   => $req->name,
                    ChartsConstants::COL_CD   => $req->code,
                    ChartsConstants::COL_DESC => $req->description,
                    ChartsConstants::COL_ENB  => $req->has(ChartsConstants::COL_ENB) ? 1 : 0,
                ]);
                $this->logExecutionTime($t, $action, 'updateAccount');

                return redirect()->route(self::REDIRECT_INDEX)->with('success', __('Account successfully updated.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action, route(self::REDIRECT_INDEX));
            }
        }, ['account_id' => $chartOfAccount->id ?? null, 'input' => $req->only('name', 'code')]);
    }

    public function destroy(Request $req, ChartOfAccount $chartOfAccount): RedirectResponse|JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($req, $chartOfAccount, $action) {
            if (($uor = self::_checkLogin()) instanceof RedirectResponse) return $uor;
            if (($c = self::guard($req, PermissionsConstants::DEL_COA, self::REDIRECT_INDEX)) !== true) return $c;

            try {
                $t = microtime(true);
                $chartOfAccount->delete();
                $this->logExecutionTime($t, $action, 'deleteAccount');

                return redirect()->route(self::REDIRECT_INDEX)->with('success', __('Account successfully deleted.'));
            } catch (\Throwable $e) {
                return defaultUndefinedException($req, $e, $action, route(self::REDIRECT_INDEX));
            }
        }, ['account_id' => $chartOfAccount->id ?? null]);
    }

    public const GET_SBT = 'getSubType';
    public function getSubType(Request $req): JsonResponse
    {
        $cls = __CLASS__;
        $meth = __METHOD__;
        $func = __FUNCTION__;
        $action = $meth;

        return $this->measureProfile($action, function () use ($req, $action) {
            try {
                $t = microtime(true);
                $types = ChartOfAccountSubType::where(ChartsConstants::COL_TP, $req->type)
                    ->pluck(ChartsConstants::COL_NM, 'id');
                $this->logExecutionTime($t, $action, 'loadSubTypes');

                return response()->json($types, Response::HTTP_OK);
            } catch (\Throwable $e) {
                return response()->json(['error' => __('An unexpected error occurred.')], Response::HTTP_INTERNAL_SERVER_ERROR);
            }
        }, ['type' => $req->input('type')]);
    }

    private static function v(Request $req, array $rules): ?RedirectResponse
    {
        $v = Validator::make($req->all(), $rules);
        return $v->fails()
            ? redirect()->back()->with('error', $v->getMessageBag()->first())
            : null;
    }

    private static function safe(
        Request  $req,
        string   $ref,
        \Closure $fn
    ): RedirectResponse|JsonResponse|View {
        try {
            return $fn();
        } catch (Throwable $e) {
            return defaultUndefinedException(
                $req,
                $e,
                __CLASS__ . '::' . $ref,
                route(self::REDIRECT_INDEX)
            );
        }
    }
}
